### ISSUE TITLE
Dynamic loading of Components

### ISSUE THREAD

**Issue by mhevery** [angular/angular#904]
# Problem

In order to support lazy loading of components or to select components based on runtime value, we need a way of selecting the component at runtime (rather than at template compile time)

## Mental Model

The place where the component should be loaded will contain headless component. (i.e. a component which does not have a `@Template`) The headless component will inject `PrivateComponent` API which will allow the headless component to bootstrap a child Component at its location. The private component will bring its own `@Template` which will serve as Shadow DOM View.

The private component is just a regular component with `@Component` and `@Template` annotation. It is in no way special and can be used in other locations normally. 

The headless component is created synchronously, but the private component is created asynchronously. The headless component can act as proxy for the private component. Only the headless component instance can be injected in the light DOM. Only the private component can be injected in the Shadow DOM.

## Example

```
@Component({
  selector: 'car',
  bind: {...},
})
@Template({
  url: '...'
})
class Car {}



@Component({
  selector: 'vehicle',
  bind: {...},
})
class Vehicle {
  constructor(pc: PrivateComponent) {
    this.pc = pc;
  }
  loadCar: () {
    someLibrary.load().then(() {
      var childInjector = pc.load(someLibrary.Car);
    })
  }
}
```

## Open Questions

 - [ ] How can we merge the `PrivateComponent` API with `bootstrap`

**Comment by mhevery** [comment]
@yjbanov @vicb @rkirov Can you comment please, so that we can iterate on this?

**Comment by rkirov** [comment]
Conceptually, I like the idea very much. We need to spec out a few more things:

* the DOM will have no reference to the private component, other than its shadow dom. In the example - ```
<vehicle> 
  #shadowRoot
     ... car template here ...
</vehicle>
```

* the component injector for `Vehicle` will contain an extra field where the `Car` instance will be stored to be accessible only for injection through the shadowDom. Note injectors do not have dynamic bindings so there would need to be a temporary injector to build `Car`, but the instance should be stored back in the injector for `Vehicle`.

Syntaxwise, I would prefer if we introduce a new directive type.
* @ViewPort is just @Decorator in which one can inject `ViewContainer`.
* @DynamicComponent is just @Component in which one can inject PrivateComponent.

With the current proposal is too easy for an new user to forget to add @Template decoration and I prefer for use to throw with a nice descriptive error `@Component annotations need at least one @Template`, versus assuming they wanted @DynamicComponent, which is most likely not the case.

**Comment by divdavem** [comment]
Hello,

I have created an **InsertComponent** directive to easily insert dynamic components from a template:
```
<div *insert="myDynamicComponent with-properties myDynamicComponentProperties"></div>
```

Each time `myDynamicComponent` changes, the component referenced by `myDynamicComponent` is compiled, then the corresponding view is instantiated, inserted and hydrated, then the given properties are assigned.

If you are interested, you can have a look to the code of the component here:
https://github.com/divdavem/ng2-play/blob/insertComponent/src/insertComponent.js
See how it can be used here:
https://github.com/divdavem/ng2-play/blob/insertComponent/src/main.js
 It is currently limited in the way properties can be set, but this could be worked on further.

If you want to play with it:
* git clone https://github.com/divdavem/ng2-play.git -b insertComponent
* npm install && gulp play

The ability to dynamically choose a component to display is a very powerful feature and could be used in many use cases (not only the router). For example, an auto-complete component could be passed as a property the component to use to display each suggestion. The template of the auto-complete would then use the **InsertComponent** directive for each suggestion...

I think a directive such as **InsertComponent** should be integrated in angular2. What do you think?

**Comment by mhevery** [comment]
Implemented: 2041860a21c021efc10589533f23c8df6dc611d3
